src/key_store/mod.rs

```diff
@@ -0,0 +1,88 @@
+use std::collections::HashMap;
+
+use crate::{
+    ciphersuite::{CiphersuiteName, SignaturePublicKey, SignatureScheme},
+    credentials::{Credential, CredentialBundle, CredentialType},
+    extensions::Extension,
+    key_packages::{KeyPackage, KeyPackageBundle},
+};
+
+pub mod errors;
+
+pub use errors::KeyStoreError;
+
+#[derive(Debug, Default)]
+/// The `KeyStore` contains private key material of `Credentials` and
+/// `KeyPackage`s.
+pub struct KeyStore {
+    // Map from signature public keys to credential bundles
+    credentials: HashMap<SignaturePublicKey, CredentialBundle>,
+    key_packages: HashMap<Vec<u8>, KeyPackageBundle>,
```

Comment:
Do the `HashMap`s really need a value for the key or could this work with a reference as well?

Reply:
No, I just meant for the key. The value has to be a value of course, but the key doesn't get stored in a `HashMap` it just gets hashed and it shouldn't be a lifetime issue.

Reply:
I see what you mean, but I'm not sure it's worth it. This would mean that the hashes of the key packages we store have to live as long as the KeyStore object. When we generate a KPB, the hash of the key package has to be computed locally. It's not part of the key package and will then go out of scope, which causes a lifetime issue. It might be possible, but I honestly just don't really see a good way to do it.

Reply:
I'm not sure the lifetime issues you describe would occur. I think `HashMap` will not take ownership here.

Reply:
Should we pair up for this? I've tried and I keep running into issues.

Reply:
It appears that the `HashMap` really needs ownership of the key that you pass in `insert`. And if you pass a reference, it needs to stay alive at least as long as the `HasMap` itself. This seems to have been a design decision. There might be a data structure that doesn't do this, but I'm not sure it would be worth using something non-standard here.

See also: https://github.com/rust-lang/rust/issues/44271